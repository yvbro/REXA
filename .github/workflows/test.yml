name: tests

on:
  push:
    branches: [ master ]
  pull_request:
    branches: [ master ]

jobs:
  build:
    runs-on: ubuntu-latest
    services:
      postgres:
        image: postgres:12-alpine
        env:
          POSTGRES_USER: rexaAdmin
          POSTGRES_PASSWORD: rexaAdmin
          POSTGRES_DB: rexa
          POSTGRES_HOST: localhost
        ports:
          - 5442:5432
        options: --health-cmd pg_isready --health-interval 10s --health-timeout 5s --health-retries 5

    steps:
      - uses: actions/checkout@v2
      - name: Set up JDK 11
        uses: actions/setup-java@v1
        with:
          java-version: 11
      - name: Init db
        run: mvn flyway:migrate
      - name: Unit test and integration test
        run: mvn -B clean verify --file pom.xml