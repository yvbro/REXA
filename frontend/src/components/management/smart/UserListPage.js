import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';

import { Button, Switch, IconButton, Grid, makeStyles } from '@material-ui/core';
import EditIcon from '@material-ui/icons/Edit';

import { switchEnabledUser } from '../redux/userDuck';
import RexaDataTable from '../../common/RexaDataTable';
import RexaModal from '../../common/RexaModal';
import RexaCard from '../../common/RexaCard';
import AddUserForm from '../smart/AddUserForm';
import ChangePasswordForm from '../smart/ChangePasswordForm';
import { GoogleIcon } from '../../auth/dumb/SocialLogin';

import { GOOGLE_AUTH_PROVIDER } from '../../../helpers/constants/index';
import themes from '../../common/theme/theme.scss';

import classes from './UserListPage.module.scss';

const DEFAULT_MODAL_PASSWORD_STATE = { open: false, userEmail: '' };

const useStyles = makeStyles(() => ({
    button: {
        marginLeft: 'auto',
        height: 40,
        marginRight: 8,
        backgroundColor: themes.primaryButtonColor,
        color: 'white',
    },
}));
const UserListPage = ({ page, setPage, rowsPerPage, setRowsPerPage }) => {
    const dispatch = useDispatch();
    const style = useStyles();

    const [openModalNewUser, setOpenModalNewUser] = useState(false);
    const [openModalPassword, setOpenModalPassword] = useState(
        DEFAULT_MODAL_PASSWORD_STATE
    );

    const { users, totalElements } = useSelector((state) => ({
        users: state.user.data,
        totalElements: state.user.totalElements,
    }));

    const handleChange = (userEmail, enabled) =>
        dispatch(switchEnabledUser(userEmail, !enabled));

    const toSwitch = (user) => {
        return (
            <Switch
                checked={user.enabled}
                onChange={() => handleChange(user.email, user.enabled)}
                disabled={user.roles.includes('ADMIN')}
                color="primary"
                name="disableUser"
                inputProps={{
                    'aria-label': 'primary checkbox',
                }}
            />
        );
    };

    const toEditComponent = (user) => {
        let component = (
            <IconButton
                color="primary"
                aria-label="edit password"
                component="span"
                onClick={() =>
                    setOpenModalPassword({ open: true, userEmail: user.email })
                }
            >
                <EditIcon />
            </IconButton>
        );

        if (user.roles.includes('ADMIN')) {
            component = (
                <IconButton
                    color="primary"
                    aria-label="edit password"
                    component="span"
                    disabled
                >
                    <EditIcon />
                </IconButton>
            );
        } else if (user.authProvider === GOOGLE_AUTH_PROVIDER) {
            component = <GoogleIcon />;
        }

        return component;
    };

    const data = [
        { name: 'Email', values: users.map((e) => e.email) },
        { name: 'Role', values: users.map((e) => e.roles.join(',')) },
        { name: 'Enabled', values: users.map((e) => toSwitch(e)) },
        { name: 'Actions', values: users.map((e) => toEditComponent(e)) },
    ];

    const action = (
        <Button
            variant="contained"
            onClick={() => setOpenModalNewUser(true)}
            className={style.button}
        >
            Add user
        </Button>
    );

    return (
        <Grid container className={classes.rootDiv} spacing={3}>
            <Grid item md={12} xs={12}>
                <RexaModal
                    open={openModalNewUser}
                    closeModal={() => setOpenModalNewUser(false)}
                    title="addUserModal"
                >
                    <div>
                        <AddUserForm
                            users={users.map((e) => e.email)}
                            closeAction={() => setOpenModalNewUser(false)}
                        />
                    </div>
                </RexaModal>
                <RexaModal
                    open={openModalPassword.open}
                    closeModal={() =>
                        setOpenModalPassword(DEFAULT_MODAL_PASSWORD_STATE)
                    }
                    title="editPasswordModal"
                >
                    <div>
                        <ChangePasswordForm
                            userEmail={openModalPassword.userEmail}
                            closeAction={() =>
                                setOpenModalPassword(DEFAULT_MODAL_PASSWORD_STATE)
                            }
                        />
                    </div>
                </RexaModal>
                <RexaCard
                    title="User Management"
                    className={classes.card}
                    actions={action}
                >
                    <RexaDataTable
                        data={data}
                        setPage={setPage}
                        setRowsPerPage={setRowsPerPage}
                        loading={false}
                        noDataLabel="No users on the platform"
                        fullHeight
                        currentPage={page}
                        totalElements={totalElements}
                        rowsPerPage={rowsPerPage}
                    />
                </RexaCard>
            </Grid>
        </Grid>
    );
};

UserListPage.propTypes = {
    page: PropTypes.number.isRequired,
    setPage: PropTypes.func.isRequired,
    rowsPerPage: PropTypes.number.isRequired,
    setRowsPerPage: PropTypes.func.isRequired,
};

export default UserListPage;
