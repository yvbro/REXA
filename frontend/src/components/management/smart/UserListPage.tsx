import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';

import { Button, Switch, IconButton, Grid, makeStyles } from '@material-ui/core';
import EditIcon from '@material-ui/icons/Edit';

import RexaDataTable from '../../common/RexaDataTable';
import RexaModal from '../../common/RexaModal';
import RexaCard from '../../common/RexaCard';
import AddUserForm from './AddUserForm';
import ChangePasswordForm from './ChangePasswordForm';
import { GoogleIcon } from '../../auth/dumb/SocialLogin';

import { GOOGLE_AUTH_PROVIDER } from '../../../helpers/constants/index';
import themes from '../../common/theme/theme.scss';

import classes from './UserListPage.module.scss';
import { UserRexa } from '../../../models/management/UserRexa';
import { Page } from '../../../models/Page';
import { RexaRole } from '../../../models/management/RexaRole';
import useUsersManagementService from '../../../services/useUsersManagementService';

const DEFAULT_MODAL_PASSWORD_STATE = { open: false, userEmail: '' };

const useStyles = makeStyles(() => ({
    button: {
        marginLeft: 'auto',
        height: 40,
        marginRight: 8,
        backgroundColor: themes.primaryButtonColor,
        color: 'white',
    },
}));

interface UserListPageProps {
    pageOfUsers: Page<UserRexa>;
    page: number;
    setPage: (page: number) => void;
    rowsPerPage: number;
    setRowsPerPage: (page: number) => void;
}

const UserListPage = ({
    page,
    setPage,
    pageOfUsers,
    rowsPerPage,
    setRowsPerPage,
}: UserListPageProps) => {
    const style = useStyles();
    const { switchEnabledUser } = useUsersManagementService();
    const [openModalNewUser, setOpenModalNewUser] = useState(false);
    const [openModalPassword, setOpenModalPassword] = useState(
        DEFAULT_MODAL_PASSWORD_STATE
    );

    const handleChange = (userEmail: string, enabled: boolean) =>
        switchEnabledUser({ userEmail: userEmail, enabled: !enabled });

    const toSwitch = (user: UserRexa) => {
        return (
            <Switch
                defaultChecked={user.enabled}
                onChange={() => handleChange(user.email, user.enabled)}
                disabled={user.roles.includes(RexaRole.ADMIN)}
                color="primary"
                name="disableUser"
                inputProps={{
                    'aria-label': 'primary checkbox',
                }}
            />
        );
    };

    const toEditComponent = (user: UserRexa) => {
        let component = (
            <IconButton
                color="primary"
                aria-label="edit password"
                component="span"
                onClick={() =>
                    setOpenModalPassword({ open: true, userEmail: user.email })
                }
            >
                <EditIcon />
            </IconButton>
        );

        if (user.roles.includes(RexaRole.ADMIN)) {
            component = (
                <IconButton
                    color="primary"
                    aria-label="edit password"
                    component="span"
                    disabled
                >
                    <EditIcon />
                </IconButton>
            );
        } else if (user.authProvider === GOOGLE_AUTH_PROVIDER) {
            component = <GoogleIcon />;
        }

        return component;
    };

    const data = [
        { name: 'Email', values: pageOfUsers.content.map((user) => user.email) },
        {
            name: 'Role',
            values: pageOfUsers.content.map((user) => user.roles.join(',')),
        },
        {
            name: 'Enabled',
            values: pageOfUsers.content.map((user) => toSwitch(user)),
        },
        {
            name: 'Actions',
            values: pageOfUsers.content.map((user) => toEditComponent(user)),
        },
    ];

    const action = (
        <Button
            variant="contained"
            onClick={() => setOpenModalNewUser(true)}
            className={style.button}
        >
            Add user
        </Button>
    );

    return (
        <Grid container className={classes.rootDiv} spacing={3}>
            <Grid item md={12} xs={12}>
                <RexaModal
                    open={openModalNewUser}
                    closeModal={() => setOpenModalNewUser(false)}
                    title="addUserModal"
                >
                    <div>
                        <AddUserForm
                            users={pageOfUsers.content.map((user) => user.email)}
                            closeAction={() => setOpenModalNewUser(false)}
                        />
                    </div>
                </RexaModal>
                <RexaModal
                    open={openModalPassword.open}
                    closeModal={() =>
                        setOpenModalPassword(DEFAULT_MODAL_PASSWORD_STATE)
                    }
                    title="editPasswordModal"
                >
                    <div>
                        <ChangePasswordForm
                            userEmail={openModalPassword.userEmail}
                            closeAction={() =>
                                setOpenModalPassword(DEFAULT_MODAL_PASSWORD_STATE)
                            }
                        />
                    </div>
                </RexaModal>
                <RexaCard
                    title="User Management"
                    className={classes.card}
                    actions={action}
                >
                    <RexaDataTable
                        data={data}
                        setPage={setPage}
                        setRowsPerPage={setRowsPerPage}
                        loading={false}
                        noDataLabel="No users on the platform"
                        fullHeight
                        currentPage={page}
                        totalElements={pageOfUsers.totalElements}
                        rowsPerPage={rowsPerPage}
                    />
                </RexaCard>
            </Grid>
        </Grid>
    );
};

export default UserListPage;
