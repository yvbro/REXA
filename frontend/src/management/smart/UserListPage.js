import React from 'react';

import {
    makeStyles,
    Table,
    TableCell,
    TableContainer,
    TableHead,
    TableBody,
    TableRow,
    withStyles,
    Paper,
    Typography,
    Switch,
} from '@material-ui/core';

import { NoUserData } from '../dumb/NoUserData';
import { switchEnabledUser } from '../redux/userDuck';
import { useDispatch, useSelector } from 'react-redux';

const StyledTableCell = withStyles((theme) => ({
    head: {
        backgroundColor: '#2b78e3',
        color: theme.palette.common.white,
    },
    body: {
        fontSize: 14,
    },
}))(TableCell);

const StyledTableRow = withStyles((theme) => ({
    root: {
        '&:nth-of-type(odd)': {
            backgroundColor: theme.palette.action.hover,
        },
    },
}))(TableRow);

const useStyles = makeStyles((theme) => ({
    root: {
        padding: theme.spacing(3),
    },
    table: {
        minWidth: 700,
    },
    header: {
        textAlign: 'center',
        background: '#f1f0eb',
    },
}));

const UserListPage = () => {
    const classes = useStyles();

    const dispatch = useDispatch();

    const { users } = useSelector((state) => ({
        users: state.user.users.data,
    }));

    const handleChange = (userEmail, enabled) =>
        dispatch(switchEnabledUser(userEmail, !enabled));

    return (
        <>
            {users ? (
                <div className={classes.root}>
                    <Typography className={classes.header} variant="h5">
                        User Management
                    </Typography>
                    <TableContainer component={Paper}>
                        <Table
                            className={classes.table}
                            aria-label="customized table"
                        >
                            <TableHead>
                                <TableRow>
                                    <StyledTableCell>Email</StyledTableCell>
                                    <StyledTableCell>Roles</StyledTableCell>
                                    <StyledTableCell>Enabled</StyledTableCell>
                                </TableRow>
                            </TableHead>
                            <TableBody>
                                {users.map((user, index) => (
                                    <StyledTableRow key={`user_${index}`}>
                                        <StyledTableCell component="th" scope="row">
                                            {user.email}
                                        </StyledTableCell>
                                        <StyledTableCell>
                                            {user.roles.join(',')}
                                        </StyledTableCell>
                                        <StyledTableCell>
                                            <Switch
                                                checked={user.enabled}
                                                onChange={() =>
                                                    handleChange(
                                                        user.email,
                                                        user.enabled
                                                    )
                                                }
                                                disabled={user.roles.includes(
                                                    'ADMIN'
                                                )}
                                                color="primary"
                                                name="disableUser"
                                                inputProps={{
                                                    'aria-label': 'primary checkbox',
                                                }}
                                            />
                                        </StyledTableCell>
                                    </StyledTableRow>
                                ))}
                            </TableBody>
                        </Table>
                    </TableContainer>
                </div>
            ) : (
                <NoUserData />
            )}
        </>
    );
};

export default UserListPage;
