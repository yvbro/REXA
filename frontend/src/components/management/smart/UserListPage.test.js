import React from 'react';
import { render, cleanup, fireEvent } from '../../../helpers/test/test-utils';
import '@testing-library/jest-dom/extend-expect';
import * as ReactRedux from 'react-redux';

import UserListPage from "./UserListPage";

const TEST_USERS = [
    { email: 'admin@test.com', roles: ['ADMIN', 'USER'], enabled: true},
    { email: 'user@test.com', roles: ['USER'], enabled: true},
    { email: 'disabled@test.com', roles: ['USER'], enabled: false},
];

const REGULAR_STATE = {
    data: TEST_USERS,
};

afterEach(cleanup)

test('should take a snapshot', () => {
    const mockXXXFn = jest.fn();
    const spyOnUseDispatch = jest
      .spyOn(ReactRedux, 'useDispatch')
      .mockReturnValue(mockXXXFn);

    const { asFragment } = render(<UserListPage />);

    expect(asFragment(<UserListPage />)).toMatchSnapshot();
})

test('should display user table properly', () => {
    const mockXXXFn = jest.fn();
    const spyOnUseDispatch = jest
      .spyOn(ReactRedux, 'useDispatch')
      .mockReturnValue(mockXXXFn);

    const { getByRole, getAllByRole } = render(<UserListPage />, {initialState: {user: REGULAR_STATE}});

    expect(getByRole('heading')).toHaveTextContent('User Management');
    expect(getByRole('table')).toBeInTheDocument();
    expect(getByRole('row', { name: 'admin@test.com ADMIN,USER primary checkbox'})).toBeInTheDocument();
    expect(getByRole('row', { name: 'user@test.com USER primary checkbox'})).toBeInTheDocument();
    expect(getByRole('row', { name: 'disabled@test.com USER primary checkbox'})).toBeInTheDocument();
})

test('should disabled checkbox if admin', () => {
    const mockXXXFn = jest.fn();
    const spyOnUseDispatch = jest
      .spyOn(ReactRedux, 'useDispatch')
      .mockReturnValue(mockXXXFn);

    const { getAllByRole } = render(<UserListPage />, {initialState: {user: REGULAR_STATE}});

    expect(getAllByRole('checkbox')[0]).toBeDisabled();
})


test('should enabled checkbox if not admin', () => {
    const mockXXXFn = jest.fn();
    const spyOnUseDispatch = jest
      .spyOn(ReactRedux, 'useDispatch')
      .mockReturnValue(mockXXXFn);

    const { getAllByRole } = render(<UserListPage />, {initialState: {user: REGULAR_STATE}});

    expect(getAllByRole('checkbox')[1]).not.toBeDisabled();
    expect(getAllByRole('checkbox')[2]).not.toBeDisabled();
})

test('should checked checkbox if user enabled', () => {
    const mockXXXFn = jest.fn();
    const spyOnUseDispatch = jest
      .spyOn(ReactRedux, 'useDispatch')
      .mockReturnValue(mockXXXFn);

    const { getAllByRole } = render(<UserListPage />, {initialState: {user: REGULAR_STATE}});

    expect(getAllByRole('checkbox')[0]).toHaveAttribute('checked');
})

test('should checked checkbox if user disabled', () => {
    const mockXXXFn = jest.fn();
    const spyOnUseDispatch = jest
      .spyOn(ReactRedux, 'useDispatch')
      .mockReturnValue(mockXXXFn);

    const { getAllByRole } = render(<UserListPage />, {initialState: {user: REGULAR_STATE}});

    expect(getAllByRole('checkbox')[2]).not.toHaveAttribute('checked');
})

// test('should checked checkbox if click on it when user disabled', () => {
//     const mockXXXFn = jest.fn();
//     const spyOnUseDispatch = jest
//       .spyOn(ReactRedux, 'useDispatch')
//       .mockReturnValue(mockXXXFn);
//     const spyOnSwitchEnabledUser = jest
//       .spyOn('../redux/userDuck', 'switchEnabledUser')
//       .mockReturnValue(mockXXXFn);

//     const { getAllByRole } = render(<UserListPage />, {initialState: {user: REGULAR_STATE}});

//     const checkbox = getAllByRole('checkbox')[2];
//     expect(checkbox).not.toHaveAttribute('checked');

//     fireEvent.click(checkbox);

//     expect(spyOnSwitchEnabledUser).toHaveBeenCalled();
//     expect(getAllByRole('test')[2]).toBeInTheDocument();
//     // expect(getAllByRole('checkbox')[2]).toHaveAttribute('checked');
// })
