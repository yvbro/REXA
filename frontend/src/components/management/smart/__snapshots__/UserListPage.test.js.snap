// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="rootDiv"
  >
    <div
      class="Header"
    >
      <h3>
        User Management
      </h3>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Add user
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiPaper-root MuiTableContainer-root makeStyles-root100-2 MuiPaper-elevation1 MuiPaper-rounded"
    >
      <table
        aria-label="sticky table"
        class="MuiTable-root MuiTable-stickyHeader"
      >
        <thead
          class="MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-4 MuiTableCell-alignCenter MuiTableCell-stickyHeader"
              scope="col"
            >
              Email
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-4 MuiTableCell-alignCenter MuiTableCell-stickyHeader"
              scope="col"
            >
              Role
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-4 MuiTableCell-alignCenter MuiTableCell-stickyHeader"
              scope="col"
            >
              Enabled
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root"
        />
      </table>
      <div
        class="MuiPaper-root MuiCard-root MuiPaper-outlined MuiPaper-rounded"
      >
        <div
          class="MuiCardContent-root makeStyles-alignItemsAndJustifyContent-7"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root makeStyles-iconDef-8"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
            />
          </svg>
          <h2
            class="MuiTypography-root MuiTypography-h5"
          >
            No users on the platform!
          </h2>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
